### PR TITLE
Restructure diffusion subpackage

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

## Restructure Diffusion Subpackage

This PR establishes the foundational structure for the `physicsnemo.diffusion` subpackage, organizing diffusion-related functionality into a clean, layered architecture.

### Summary

- **Created `physicsnemo.diffusion` package** with dedicated submodules:
  - `generate/` - Higher-level generation utilities and pipelines
  - `samplers/` - Samplers based on ODE and SDE solvers
  - `metrics/` - Loss functions and diffusion-specific metrics (FID, etc.)
  - `preconditioners/` - Preconditioning model wrappers
  - `multi_diffusion/` - Patch-based diffusion utilities
  - `denoisers/` - Placeholder for future denoiser implementations; to be used in samplers
  - `noise_schedulers/` - Placeholder for future noise scheduler implementations
  - `utils/` - Shared utilities 

- **Added warning system for incremental migration**:
  - `FutureFeatureWarning` for placeholder modules not yet implemented
  - `LegacyFeatureWarning` for existing code that will be deprecated in future releases

- **Reorganized UNet models**:
  - Moved `SongUNet` models and application-specific UNet wrappers to `physicsnemo.models.diffusion_unets`

- **Extracted reusable layers into `physicsnemo.nn`**:
This includes for example:
  - `group_norm.py` - GroupNorm variants with optional FP32 conversion
  - `unet_layers.py` - UNet encoder/decoder blocks, attention blocks
  - `embedding_layers.py` - Positional and Fourier embeddings
Etc...

- **Cleaned up legacy code**:
  - Deleted unused utility functions from old `physicsnemo.models.diffusion/`
  - Moved example-specific utilities directly into `examples/` directory

- **Added import-linter contracts** for the diffusion package:
  - Enforces layered dependency structure: `generate` → `samplers`/`metrics` → `preconditioners`/`multi_diffusion`/`denoisers`/`noise_schedulers` → `utils`
  - Prevents circular imports and maintains clean architecture

- **Updated all examples and tests** to use the new package structure

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: docs/api/models/diffusion.rst
Comment (pzharrington) [inline docs/api/models/diffusion.rst:1]: The references throughout this file (in the text) need to be updated, not just those that are part of the rendered code blocks

File: examples/cfd/flow_reconstruction_diffusion/dataset/dataset.py [hunk 1]
```diff
@@ -23,7 +23,7 @@
 import numpy as np
 import PIL.Image
 import torch
-from physicsnemo.models.diffusion.training_utils import EasyDict
+from misc import EasyDict
```
Comment (pzharrington) [inline examples/cfd/flow_reconstruction_diffusion/dataset/dataset.py:26]: Does this import work? `misc` is in a directory above this file

File: examples/cfd/flow_reconstruction_diffusion/training_loop.py [hunk 2]
```diff
@@ -45,6 +45,22 @@
 # ----------------------------------------------------------------------------
```
Comment (pzharrington) [inline examples/cfd/flow_reconstruction_diffusion/training_loop.py]: LOL ☝️

File: physicsnemo/compat/__init__.py
Comment (pzharrington) [inline physicsnemo/compat/__init__.py:1]: Did this somehow get removed from Corey's v2.0 PR, and you are reintroducing it here?

File: physicsnemo/diffusion/multi_diffusion/multi_diffusion.py
Comment (pzharrington) [inline physicsnemo/diffusion/multi_diffusion/multi_diffusion.py:1]: What is `physicsnemo.diffusion.multi_diffusion.multi_diffusion` supposed to add in addition to what is already now exposed here under `multi_diffusion.patching`?

Comment (pzharrington) [reply physicsnemo/diffusion/multi_diffusion/multi_diffusion.py:1]: (Since I don't see deprecation warnings on the patching stuff I assume it is here to stay?

Comment (CharlelieLrt) [reply physicsnemo/diffusion/multi_diffusion/multi_diffusion.py:1]: My plan is to have some high-level multi-diffusion model wrapper. Ideally, something like:

```python
model = SongUNet(...)  # Model architecture
multi_diffusion_model = MultiDiffusionModel(
    model=model,
    position_embeddings="learnable"
    pos_embd_channels=20
)
x = torch.randn(4, 8, 2000, 2000) # Some very large input
t = torch.rand((4,))
condition = torch.randn(4, 8)
x0 = multi_diffusion_model(x, t, condition)  # Directly pass the very large input without patching
```
And the `MultiDiffusionModel` internally uses the functionalities currently exposed through `patching` to patch the input, extract the corresponding positional embeddings, (optionally) fuse the resulting patches, etc...

In the current implementation all the patching operations are carried out in the samplers, but this is not ideal. Samplers should be as simple as an ODE/SDE numerical integration method (e.g. Euler, Heun), combined with a noise schedule. They should be totally agnostic to patching. So, the plan is to move all the patching ops to this `MultiDiffusionWrapper`.

That's my plan at least, happy to discuss more about it.

File: physicsnemo/models/diffusion_unets/_utils.py [hunk 3]
```diff
@@ -17,56 +17,6 @@
...
-import numpy as np
-import torch
-
-
-def weight_init(shape: tuple, mode: str, fan_in: int, fan_out: int):
-    """
-    Unified routine for initializing weights and biases.
-    This function provides a unified interface for various weight initialization
-    strategies like Xavier (Glorot) and Kaiming (He) initializations.
-
```
Comment (pzharrington) [inline physicsnemo/models/diffusion_unets/_utils.py:29]: I may have missed it...did this function get moved somewhere else? Or is it unused?

File: physicsnemo/compat/__init__.py
Comment (CharlelieLrt) [reply physicsnemo/compat/__init__.py:1]: The git merge was an absolute mess, so it's very possible I made mistake...
But after merging with Corey's v2.0PR what I got here was an empty file with an unused `import os`. I assume Corey just forgot to remove it? So I deleted the unused import
@coreyjadams for viz

File: physicsnemo/nn/attention_layers.py [hunk 4]
```diff
@@ -14,12 +14,175 @@
...
+        dk = torch.einsum("ncq,nqk->nck", q.to(torch.float32), db).to(
+            k.dtype
+        ) / np.sqrt(k.shape[1])
+        return dq, dk
+
+
+class Attention(torch.nn.Module):
+    """
+    Self-attention block used in U-Net-style architectures, such as DDPM++, NCSN++, and ADM.
+    Applies GroupNorm followed by multi-head self-attention and a projection layer.
```
Comment (pzharrington) [inline physicsnemo/nn/attention_layers.py:72]: I don't love having the module named `nn.attention_layers.Attention` be this relatively model-specific U-Net version -- seems like such a module should be relatively general. Thoughts on renaming it?

File: physicsnemo/diffusion/multi_diffusion/multi_diffusion.py
Comment (CharlelieLrt) [reply physicsnemo/diffusion/multi_diffusion/multi_diffusion.py:1]: > (Since I don't see deprecation warnings on the patching stuff I assume it is here to stay?


Yes. Both exposed to the user and used internally by `MultiDiffusionModel`

File: physicsnemo/nn/utils/utils.py
Comment (pzharrington) [inline physicsnemo/nn/utils/utils.py:1]: Probably more of a nit for @loliverhennigh and the general `nn` refactor, but `nn.utils.utils` is a bit messy. We should probably just consolidate into one private `_utils.py` file?

File: physicsnemo/models/diffusion_unets/_utils.py (hunk 3 above)
Comment (pzharrington) [reply physicsnemo/models/diffusion_unets/_utils.py:29]: Nvm, I see it's moved to `nn` now

File: physicsnemo/nn/utils/utils.py
Comment (pzharrington) [reply physicsnemo/nn/utils/utils.py:1]: Could be public too I guess if we think users would like to import them

File: physicsnemo/compat/__init__.py
Comment (pzharrington) [reply physicsnemo/compat/__init__.py:1]: Well I don't see a `physicsnemo.compat` in `main` branch at the moment so I think it was somehow dropped from the big v2.0 PR?

Comment (coreyjadams) [reply physicsnemo/compat/__init__.py:1]: Yeah my rebase onto main after the merge was also  a disaster.  I think squashing commits makes it impossible for git to actually merge properly, and we had too many squashed?

I had removed compat from v2.0 - I thought it was breaking doc tests (I no longer think this) but I also thought, "is this really useful?".  Anyways it sounds like there is more will to reintroduce it, I won't stand in the way.

File: physicsnemo/diffusion/multi_diffusion/multi_diffusion.py
Comment (pzharrington) [reply physicsnemo/diffusion/multi_diffusion/multi_diffusion.py:1]: Got it. I like the high level plan of being able to wrap backbones so they can do patched diffusion, I think it's critical they have basically the same interface for calling during training vs. sampling (which allows you to keep the samplers "patch-naive" and simple which is the right move). In theory then the user would be able to easily choose between domain parallelism or patching (depending on if their GPU count 😅 is big or small) without too much code changes

File: physicsnemo/compat/__init__.py
Comment (pzharrington) [reply physicsnemo/compat/__init__.py:1]: I think it's def useful as it'll be another tool in our belt for helping people migrate

Comment (coreyjadams) [reply physicsnemo/compat/__init__.py:1]: OK, I'll bring it back in a standalone PR.  @CharlelieLrt is innocent in this one!

File: physicsnemo/nn/utils/utils.py
Comment (CharlelieLrt) [reply physicsnemo/nn/utils/utils.py:1]: Agree. Just to be clear, the `_validate_amp` that I added there is to be deprecated in a future PR. Just moved it there because I could not find a better place for it

**Comment by megnvidia** [comment]
I had small changes to the two rst files. If those get merged well, I approve.

File: docs/api/models/diffusion.rst
Comment (CharlelieLrt) [reply docs/api/models/diffusion.rst:1]: Done

File: examples/cfd/flow_reconstruction_diffusion/dataset/dataset.py (hunk 1 above)
Comment (CharlelieLrt) [reply examples/cfd/flow_reconstruction_diffusion/dataset/dataset.py:26]: This example handles imports in a very strange way, but I think it does work. My IDE can find it at least

File: physicsnemo/compat/__init__.py
Comment (CharlelieLrt) [reply physicsnemo/compat/__init__.py:1]: I will remove the `physicsnemo/compat` in this PR. @coreyjadams feel free to bring it back in a standalone PR. I agree with @pzharrington: it's a useful feature, and we could even leverage it in `Module.from_checkpoint` to seamlessly load checkpoints generated with 1.x.x versions of physicsnemo (might be useful for E2S)

File: examples/cfd/flow_reconstruction_diffusion/training_loop.py (hunk 2 above)
Comment (CharlelieLrt) [reply examples/cfd/flow_reconstruction_diffusion/training_loop.py]: Deleted 😅

File: physicsnemo/nn/attention_layers.py (hunk 4 above)
Comment (CharlelieLrt) [reply physicsnemo/nn/attention_layers.py:72]: Renamed `Attention` to `UNetAttention`
